staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -711,28 +774,44 @@ func (c *Cacher) startDispatching(event *watchCacheEvent) {
 	// gain from avoiding memory allocations is much bigger.
 	c.watchersBuffer = c.watchersBuffer[:0]
 
-	// Iterate over "allWatchers" no matter what the trigger function is.
-	for _, watcher := range c.watchers.allWatchers {
-		c.watchersBuffer = append(c.watchersBuffer, watcher)
-	}
-	if supported {
-		// Iterate over watchers interested in the given values of the trigger.
-		for _, triggerValue := range triggerValues {
-			for _, watcher := range c.watchers.valueWatchers[triggerValue] {
+	if event.Type == watch.Bookmark {
+		// Pop up watchers from startBucketID to endBucketID, ignore those already
+		// stopped as we do not delete the watcher when it is stopped,
+		for bucketID := startBucketID; bucketID < endBucketID; bucketID++ {
+			for _, watcher := range c.bookmarkWatchers.popWatchers(bucketID) {
+				if watcher.stopped {
+					continue
+				}
```

Comment:
This works, because crrently cacheWatcher.stop() is always called when cacher.Lock() is held.
I think it's fine, but it requires adding an explicit comment both here and for the cacheWatcher.stop() method so that it won't be changed by mistake later.

Reply:
Actually cacheWatcher.stop() is protected by c.Lock(), we do not need a sync.Mutex in cacheWatcher. I will remove it in another PR, adding a TODO there.

Reply:
Since last 2 weeks or so - yes. But I'm not entirely sure if that's the best option and if we can't reduce contention by changing the pattern a bit more. Let's leave any changes for later.